text/0045-add-notion-of-protocol.md

```diff
@@ -0,0 +1,146 @@
+# Summary
+- Define supported protocols for services and outputs in libobs and OBS Studio
+- Outputs for streaming are registered with one or more compatible protocols and codecs
+- Only codecs compatible with the output and the service are shown
+- Audio codec can be chosen if various compatible, codec in simple mode and encoder advanced output mode.
+- Will be extended with RFC 39
+
+# Motivation
+Create a better management of outputs, encoders, and services with an approach focused on protocols.
+
+# Design
+
+## Outputs API
+Only outputs with the `OBS_OUTPUT_SERVICE` flag are concerned.
+
+Adding to `obs_output_info`:
+- `const char *protocols`: protocols separated with a semi-colon (ex: `"RTMP;RTMPS"`), required to register the output.
+- `const char *protocols_prefixes`: URL protocols prefixes separated with a semi-colon `"rtmp://;rtmps://"`, optional.
+  - Up to one URL prefix per protocol in `protocols`.
+  - Those URL prefixes must be in the same order as `protocols`.
+  - Each URL prefix must be unique to the protocol
+  - Meant to be used for protocol auto-detection on server URL.
+  - While registering if present, the number of URL prefix must be less or equal to the number of protocols.
+  - NOTE: This field is optional because not every protocol has a unique prefix.
+  - NOTE 2: URL prefixes (URI scheme + `://`) are prefered over URI schemes to avoid:
+    - matching links without prefixes (e.g., "rtmp.example.com").
+    - adding `://` each time that a auto-detection is done.
+
+Adding to this API these functions:
+- `const char *obs_output_get_protocols(const obs_output_t *output)`: returns protocols supported by the output.
+- `bool obs_output_is_protocol_registered(const char *protocol)`: return true if an output with the protocol is registered.
```

Comment:
Could you explain (preferably with an example) why such a function is useful ? Seems to me that internal plugins won't ever use an unimplemented protocol. Regarding third-party plugins (for a service for example), why would they use an unimplemented protocol ? If for instance a plugin 1 (like obs-ndi) is required by plugin 2, it's up to plugin 2 to make sure plugin 1 is installed.

Reply:
It may happen with internal plugins, we might support in the future protocol that rely on library that are not on LTS Linux distro.

> Regarding third-party plugins (for a service for example), why would they use an unimplemented protocol ?

This requires RFC 39, for the same reason about future plugins or because the service might support a protocol that comes from another plugin.

> If for instance a plugin 1 (like obs-ndi) is required by plugin 2, it's up to plugin 2 to make sure plugin 1 is installed.

It's impossible while loading plugins, plugin 2 can be loaded before plugin 1. And so even if plugin 1 is there, plugin 2 can't see it is.

---

This function just enables improving UI/UX, not showing services or not listing protocol in services property view that will not work, by asking to the API if there is a registered output that support this protocol.

Reply:
thanks for the clarifications